### PR TITLE
vulkaninfo: dynamically load vulkan

### PR DESCRIPTION
**Pull request by charles-lunarg**
This commit makes it so that vulkaninfo will not statically link to the vulkan
loader dll. This is to satisfy the Microsoft ApiValidator requirements.

Change-Id: Ibac46dd13df9568e9a97ce1ba4e50613902848a1

### CODE REVIEW

File: vulkaninfo/vulkaninfo.h [hunk 1]
```diff
@@ -230,7 +235,224 @@ auto GetVector(F &&f, Ts &&... ts) -> std::vector<T> {
...
+    template <typename T>
+    void load(T &func_dest, const char *func_name) {
+#if defined(__linux__) || defined(__APPLE__)
+        func_dest = reinterpret_cast<T>(dlsym(library, func_name));
+#elif defined(_WIN32)
+        func_dest = reinterpret_cast<T>(GetProcAddress(library, func_name));
+#endif
+    }
+#if defined(__linux__) || defined(__APPLE__)
+    void *library;
```
Comment (mikew-lunarg) [inline vulkaninfo/vulkaninfo.h]: Per [cppguide](https://google.github.io/styleguide/cppguide.html#Variable_Names), class data members have "_" appended.

File: vulkaninfo/vulkaninfo.h [hunk 2]
```diff
@@ -230,7 +235,224 @@ auto GetVector(F &&f, Ts &&... ts) -> std::vector<T> {
...
+        if (library == 0) return VK_ERROR_INITIALIZATION_FAILED;
+        return VK_SUCCESS;
+    }
+    void close() {
+#if defined(__linux__) || defined(__APPLE__)
+        dlclose(library);
+#elif defined(_WIN32)
+        FreeLibrary(library);
+#endif
+    }
```
Comment (mikew-lunarg) [inline vulkaninfo/vulkaninfo.h]: suggest adding "library = 0;" here

File: vulkaninfo/vulkaninfo.h [hunk 3]
```diff
@@ -230,7 +235,224 @@ auto GetVector(F &&f, Ts &&... ts) -> std::vector<T> {
...
+#ifdef VK_USE_PLATFORM_WIN32_KHR
+    PFN_vkGetPhysicalDeviceWin32PresentationSupportKHR vkGetPhysicalDeviceWin32PresentationSupportKHR;
+#endif  // VK_USE_PLATFORM_WIN32_KHR
+#ifdef VK_USE_PLATFORM_MACOS_MVK
+    PFN_vkCreateMacOSSurfaceMVK vkCreateMacOSSurfaceMVK;
+#endif  // VK_USE_PLATFORM_MACOS_MVK
+#ifdef VK_USE_PLATFORM_METAL_EXT
+    PFN_vkCreateMetalSurfaceEXT vkCreateMetalSurfaceEXT;
+#endif  // VK_USE_PLATFORM_METAL_EXT
+    void init_dispatch_pointers() {
```
Comment (mikew-lunarg) [inline vulkaninfo/vulkaninfo.h]: method name should be lower-camelcase?

File: vulkaninfo/vulkaninfo.h [hunk 4]
```diff
@@ -353,13 +564,17 @@ struct AppInstance {
     ANativeWindow *window;
 #endif
     AppInstance() {
-        PFN_vkEnumerateInstanceVersion enumerate_instance_version =
-            reinterpret_cast<PFN_vkEnumerateInstanceVersion>(vkGetInstanceProcAddr(nullptr, "vkEnumerateInstanceVersion"));
+        if (dll.initialize() != VK_SUCCESS) {
+            fprintf(stderr, "Failed to initialize: Vulkan loader is not installed, not found or failed to load.\n");
+            WAIT_FOR_CONSOLE_DESTROY;
+            exit(1);
```
Comment (mikew-lunarg) [inline vulkaninfo/vulkaninfo.h]: ERR_EXIT() for consistency with below?

Comment (charles-lunarg) [reply vulkaninfo/vulkaninfo.h]: ERR_EXIT doesn't work here, as this isn't a Vulkan function call failing.

File: vulkaninfo/vulkaninfo.h (hunk 2 above)
Comment (charles-lunarg) [reply vulkaninfo/vulkaninfo.h]: Added.

File: vulkaninfo/vulkaninfo.h (hunk 3 above)
Comment (charles-lunarg) [reply vulkaninfo/vulkaninfo.h]: Made the functions consistent with the rest of vulkaninfo, which is PascalCase.

File: vulkaninfo/vulkaninfo.h (hunk 4 above)
Comment (mikew-lunarg) [reply vulkaninfo/vulkaninfo.h]: ah, didn't realize it was Vulkan-specific error reporting

**Pull request merged**